pkg/recommender/engine.go

```diff
@@ -18,34 +18,102 @@ import (
 
 // NewEngine creates a new Engine instance
-func NewEngine(nodePools NodePoolRecommender, log logur.Logger) *Engine {
+func NewEngine(log logur.Logger, vms VmRecommender, nodePools NodePoolRecommender, ciSource CloudInfoSource) *Engine {
 	return &Engine{
-		np:  nodePools,
-		log: log,
+		log:      log,
+		vms:      vms,
+		np:       nodePools,
+		ciSource: ciSource,
 	}
 }
 
 // RecommendCluster performs recommendation based on the provided arguments
-func (e *Engine) RecommendCluster(provider string, service string, region string, req ClusterRecommendationReq, layoutDesc []NodePoolDesc, log logur.Logger) (*ClusterRecommendationResp, error) {
-	e.log = log
-
+func (e *Engine) RecommendCluster(provider string, service string, region string, req ClusterRecommendationReq, layoutDesc []NodePoolDesc) (*ClusterRecommendationResp, error) {
 	e.log.Info(fmt.Sprintf("recommending cluster configuration. request: [%#v]", req))
 
-	nodePools, err := e.np.RecommendNodePools(provider, service, region, req, log, layoutDesc)
+	desiredCpu := req.SumCpu
+	desiredMem := req.SumMem
+	desiredOdPct := req.OnDemandPct
+
+	attributes := []string{Cpu, Memory}
+	nodePools := make(map[string][]NodePool, 2)
+
+	allProducts, err := e.ciSource.GetProductDetails(provider, service, region)
 	if err != nil {
 		return nil, err
 	}
 
+	if req.OnDemandPct != 100 {
+		availableSpotPrice := false
+		for _, vm := range allProducts {
+			if vm.SpotPrice != nil {
+				availableSpotPrice = true
+				break
+			}
+		}
+		if !availableSpotPrice {
+			e.log.Warn("onDemand percentage in the request ignored")
+			req.OnDemandPct = 100
+		}
+	}
+
+	for _, attr := range attributes {
+		vmsInRange, err := e.vms.FindVmsWithAttrValues(attr, req, layoutDesc, allProducts)
+		if err != nil {
+			return nil, emperror.With(err, RecommenderErrorTag, "vms")
+		}
+
+		layout := e.transformLayout(layoutDesc, vmsInRange)
+		if layout != nil {
+			req.SumCpu, req.SumMem, req.OnDemandPct, err = e.computeScaleoutResources(layout, attr, desiredCpu, desiredMem, desiredOdPct)
+			if err != nil {
+				e.log.Error(emperror.Wrap(err, "failed to compute scaleout resources").Error())
+				continue
+			}
+			if req.SumCpu < 0 && req.SumMem < 0 {
+				return nil, emperror.With(fmt.Errorf("there's already enough resources in the cluster. Total resources available: CPU: %v, Mem: %v", desiredCpu-req.SumCpu, desiredMem-req.SumMem))
+			}
+		}
+
+		odVms, spotVms, err := e.vms.RecommendVms(provider, vmsInRange, attr, req, layout)
+		if err != nil {
+			return nil, emperror.Wrap(err, "failed to recommend virtual machines")
+		}
+
+		if (len(odVms) == 0 && req.OnDemandPct > 0) || (len(spotVms) == 0 && req.OnDemandPct < 100) {
+			e.log.Debug("no vms with the requested resources found", map[string]interface{}{"attribute": attr})
+			// skip the nodepool creation, go to the next attr
+			continue
+		}
+		e.log.Debug("recommended on-demand vms", map[string]interface{}{"attribute": attr, "count": len(odVms), "values": odVms})
```

Comment:
this is duplicated information here; don't debug in 2 consecutive lines; rather do l;og eher the informaiton to be logged becomes available.